src/job_manager.ts

```diff
@@ -0,0 +1,160 @@
+        this.launcherQueue = this.createQueue(JobManager.launcherQueueName, options.queueRedisOptions, (ctx, group) =>
+            this.instanceLauncher.launchOrShutdownInstancesByGroup(ctx, group),
+        );
+    }
+
+    createQueue(
+        queueName: string,
+        redisClientOptions: ClientOpts,
+        processingHandler: (ctx: context.Context, group: InstanceGroup) => Promise<boolean>,
+    ): Queue {
+        const newQueue = new Queue(queueName, {
+            redis: redisClientOptions,
+            removeOnSuccess: true,
+            removeOnFailure: true,
+        });
+        newQueue.on('error', (err) => {
+            logger.error(`A queue error happened in queue ${queueName}: ${err.message}`);
+        });
+        newQueue.on('failed', (job, err) => {
+            logger.error(`Job ${job.id} in queue ${queueName} failed with error ${err.message}`);
+        });
+
+        newQueue.process((job: Job, done: DoneCallback<boolean>) => {
+            const start = Date.now();
+            const pollId = shortid.generate();
+            const pollLogger = logger.child({
+                id: pollId,
+            });
+            const ctx = new context.Context(pollLogger, start, pollId);
+            ctx.logger.info(
+                `[QueueProcessor] Start processing job ${job.id} in queue ${queueName} for group ${job.data.name}`,
+            );
+            processingHandler(ctx, job.data)
+                .then((result: boolean) => {
+                    ctx.logger.info(
+                        `[QueueProcessor] Done processing job ${job.id} in queue ${queueName} for group ${job.data.name}`,
+                    );
+                    return done(null, result);
+                })
+                .catch((error: any) => {
+                    ctx.logger.info(
+                        `[QueueProcessor] Error processing job ${job.id} in queue ${queueName} for group ${job.data.name}: ${error}`,
+                        {
+                            jobId: job.id,
+                            jobData: job.data,
+                        },
+                    );
+                    return done(error, false);
+                });
+        });
+
+        return newQueue;
+    }
+
+    async createGroupProcessingJobs(ctx: context.Context): Promise<void> {
+        if (!(await this.instanceGroupManager.isGroupJobsCreationAllowed())) {
+            ctx.logger.info('[JobManager] Job creation not allowed yet, nothing to be done');
+            return;
+        }
+
+        let lock: Redlock.Lock = undefined;
+        try {
+            lock = await this.lockManager.lockJobCreation(ctx);
+        } catch (err) {
+            ctx.logger.warn(`[JobManager] Error obtaining lock for creating jobs`, { err });
+            return;
+        }
+
+        try {
+            if (!(await this.instanceGroupManager.isGroupJobsCreationAllowed())) {
+                ctx.logger.info('[JobManager] Job creation not allowed yet, nothing to be done');
```

Comment:
Shouldn't this unlock before returning?

Reply:
The unlock is in the finally block and it should always happen, even if we return here

Reply:
I'm not sure finally is run if we return from a function, I believe it will only run if we exit the try block normally or by error.

Reply:
Whelp I tested this and you're right!  Consider my comment incorrect!